packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx

```diff
@@ -30,53 +32,55 @@ function transformTracesToPaths(
     if (data) {
       const spanMap: Map<string, Span> = new Map();
       const { traceID } = data;
-      data.spans
-        .filter(span => {
-          spanMap.set(span.spanID, span);
-          return !span.hasChildren;
-        })
-        .forEach(leaf => {
-          const spans = spanAncestorIds(leaf).map(id => {
-            const span = spanMap.get(id);
-            if (!span) throw new Error(`Ancestor spanID ${id} not found in trace ${traceID}`);
-            return span;
-          });
-          spans.reverse();
-          spans.push(leaf);
-
-          const path: TDdgPayloadEntry[] = spans
-            .filter(span => span.tags.find(({ key, value }) => key === 'span.kind' && value === 'server'))
-            .map(({ processID, operationName: operation }) => ({
-              service: data.processes[processID].serviceName,
-              operation,
-            }));
-
-          if (
-            path.some(
-              ({ service, operation }) =>
-                service === focalService && (!focalOperation || operation === focalOperation)
-            )
+      data.spans.forEach(span => spanMap.set(span.spanID, span));
+      const tree = getTraceSpanIdsAsTree(data);
+      tree.paths((pathIds: string[]) => {
+        const paths = pathIds.reduce((reducedSpans: Span[], id: string): Span[] => {
+          if (id === '__root__') {
+            return reducedSpans;
+          }
+          const span = spanMap.get(id);
```

Comment:
Q: why does `tree.paths()` generate paths as `ids[]` instead of `spans[]`? The tree descent does operate on spans, and the paths generator does not know what the paths will be used for, so returning IDs instead of spans seems unnecessary, only requiring you to re-map IDs back to spans.

Reply:
This is a good point, and actually the reason I didn't do in that way is that I'm using the same tree used in transform-trace-data, which in each node only have the `spanID`.

We can of course change that to store the whole span on the node, but that will require other changes.

Reply:
fair enough